### PR TITLE
feat: logging support for requests and responses

### LINKED ISSUE #57: Add logging support for API requests and responses
Provide the support for following configurations

## General Logging Configuration Options

- Level: Defines the log message severity or priority (e.g., DEBUG, INFO, WARN, ERROR, FATAL).
- Logger: Allows users to provide a custom logger implementation.
- Mask Sensitive Headers: A global setting to mask sensitive HTTP headers in both requests and responses before logging, safeguarding confidential data.

### Request Logging Configuration Options
- Log Body: Controls the logging of the request body.
- Log Headers: Controls the logging of request headers.
- Exclude Headers: Excludes specified headers from the log output.
- Include Headers: Includes only specified headers in the log output.
- Unmask Headers: Logs specified headers without masking, revealing their actual values.
- Include Query in Path: Determines whether to include query parameters in the logged request path.
### Response Logging Configuration Options
- Log Body: Controls the logging of the response body.
- Log Headers: Controls the logging of response headers.
- Exclude Headers: Excludes specified headers from the log output.
- Include Headers: Includes only specified headers in the log output.
- Unmask Headers: Logs specified headers without masking, revealing their actual values.

### PR DESCRIPTION
**Pull request by asadali214**
## What
This PR adds the configurable logging support via psr/log/LoggerInterface. It adds the ability to log requests and responses for any API call. It also adds the unit tests that cover all aspects of this new feature

## Why
To improve developer experience for maintaining and debugging API logs. Read more details in the attached issue.

Closes #57 

## Type of change
Select multiple if applicable.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause a breaking change)
- [X] Tests (adds or updates tests)
- [ ] Documentation (adds or updates documentation)
- [ ] Refactor (style improvements, performance improvements, code refactoring)
- [ ] Revert (reverts a commit)
- [ ] CI/Build (adds or updates a script, change in external dependencies)

## Dependency Change
This PR adds the dependency of core-lib-php on: `psr/log` versions: `^1.1.4 || ^2.0.0 || ^3.0.0`.

## Breaking change
N/A

## Testing
Unit tests were added and 100% new code coverage has been achieved. 

## Checklist
- [X] My code follows the coding conventions
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added new unit tests

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -16,7 +16,8 @@
     "ext-curl": "*",
     "ext-dom": "*",
     "ext-libxml": "*",
-    "apimatic/core-interfaces": "~0.1.4",
+    "psr/log": "^1.1.4 || ^2.0.0 || ^3.0.0",
+    "apimatic/core-interfaces": "dev-test-logging",
```
Comment (asadali214) [inline composer.json:20]: @sufyankhanrao I will change this to `~0.1.5`, after releasing core-interfaces version, before merging this PR.

**Review by sufyankhanrao — CHANGES_REQUESTED** [review]

File: src/Logger/Configuration/LoggingConfiguration.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * Log the given message using the context array. This function uses the
+     * LogLevel and Logger instance set via constructor of this class.
+     */
+    public function logMessage(string $message, array $context): void
+    {
+        switch ($this->level) {
+            case LogLevel::DEBUG:
+                $this->logger->debug($message, $context);
```
[the remark this answers is not in the document]
Comment (asadali214) [reply src/Logger/Configuration/LoggingConfiguration.php:41]: I followed this standard: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md#11-basics.

However yes it does say `Calling log() method with one of the log level constants MUST have the same result as calling the level-specific method.`.

So I will replace this switch with `log` function call

File: src/Logger/Configuration/LoggingConfiguration.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                break;
+            case LogLevel::ALERT:
+                $this->logger->alert($message, $context);
+                break;
+            case LogLevel::EMERGENCY:
+                $this->logger->emergency($message, $context);
+                break;
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (asadali214) [reply src/Logger/Configuration/LoggingConfiguration.php:66]: These getters are exposed via SDKs

File: src/Logger/Configuration/HttpConfiguration.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        bool $logHeaders,
+        array $headersToInclude,
+        array $headersToExclude,
+        array $headersToUnmask
+    ) {
+        $this->logBody = $logBody;
+        $this->logHeaders = $logHeaders;
+        $this->headersToInclude = array_map('strtolower', $headersToInclude);
+        $this->headersToExclude = empty($headersToInclude) ? array_map('strtolower', $headersToExclude) : [];
+        $this->headersToUnmask = array_merge(
```
[the remark this answers is not in the document]
Comment (asadali214) [reply src/Logger/Configuration/HttpConfiguration.php:37]: These getters are exposed via SDKs and yes the `headersToUnmask` configuration is immutable from user's perspective as this property is only being used internally

File: src/Logger/Configuration/HttpConfiguration.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Core\Logger\Configuration;
+
+use Core\Logger\LoggerConstants;
+
+class HttpConfiguration
```
[the remark this answers is not in the document]
Comment (asadali214) [reply src/Logger/Configuration/HttpConfiguration.php]: Added suggestion

File: src/Logger/Configuration/LoggingConfiguration.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (asadali214) [reply src/Logger/Configuration/LoggingConfiguration.php]: Added suggestion

File: src/Client.php [hunk 6]
```diff
@@ -128,12 +142,18 @@ public function beforeRequest(Request $request)
         if (isset($this->apiCallback)) {
             $this->apiCallback->callOnBeforeWithConversion($request, self::getConverter($this));
         }
+        if (isset($this->apiLogger)) {
```
[the remark this answers is not in the document]
Comment (asadali214) [reply src/Client.php]: Added suggestion

File: README.md [hunk 7]
```diff
@@ -52,15 +52,15 @@ composer require "apimatic/core"
...
-| [`HttpConfiguration`](src/Logger/Configuration/HttpConfiguration.php)         | Common configurations shared by request and response logging configurations |
-| [`LoggingConfiguration`](src/Logger/Configuration/LoggingConfiguration.php)   | Provides client's logging configurations                                    |
-| [`RequestConfiguration`](src/Logger/Configuration/RequestConfiguration.php)   | Provides request's logging configurations                                   |
-| [`ResponseConfiguration`](src/Logger/Configuration/ResponseConfiguration.php) | Provides response's logging configurations                                  |
+| Name                                                                                        | Description                                                                 |
+|---------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------|
+| [`ApiLogger`](src/Logger/ApiLogger.php)                                                     | Provides implementation for logging API requests and responses              |
```
Comment (sufyankhanrao) [inline README.md:57]: Blocker: The documentation around NullApiLogger should also be added here.

**Review by sufyankhanrao — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (asadali214) [reply README.md]: resolved!

**Review by sufyankhanrao — APPROVED** [review]

**Pull request merged**